### PR TITLE
[feature] #3021: Opaque pointer import

### PR DESCRIPTION
**Pull request by mversic**
Signed-off-by: Marin Veršić <marin.versic101@gmail.com>

### Description of the Change

* add support for opaque pointer types on FFI import
* rename `FfiWrapperOutput` to `FfiWrapperType`
* fix array conversions

### CODE REVIEW

File: ffi/src/ir.rs [hunk 1]
```diff
@@ -100,57 +122,156 @@ unsafe impl<R: Transmute, const N: usize> Transmute for [R; N] {
...
 }
-
 impl<R> Ir for *mut R {
     type Type = Robust;
 }
 
-impl<'itm, R: Ir + Cloned> Ir for &'itm R {
-    type Type = &'itm R::Type;
+/// When implemented for a type, defines how dependent types are mapped into [`Ir`]
+pub trait GroupBy {
```
Comment (mversic) [inline ffi/src/ir.rs:136]: will rename to something more appropriate

File: ffi/tests/ffi_export.rs [hunk 2]
```diff
@@ -613,8 +616,7 @@ fn data_carrying_enum_conversion() {
             )
         );
 
-        let mut store = Default::default();
-        let ret_val = FfiConvert::try_from_ffi(output.assume_init(), &mut store);
+        let ret_val = FfiOutPtrRead::try_read_out(output.assume_init());
```
[the remark this answers is not in the document]
Comment (mversic) [reply ffi/tests/ffi_export.rs:619]: it's essentially the same thing. It's possible I just favored this for aesthetic reasons. Might not pertain to this in particular, but I remember there was an issue in some case with using `try_from_ffi`. Could be that it didn't work when tested for wasm architecture. Anyway, `try_read_out` is a wrapper for `try_from_ffi` so no semantic difference here.

File: ffi/tests/ffi_export.rs [hunk 3]
```diff
@@ -180,6 +180,12 @@ pub fn get_vec_of_boxed_opaques() -> Vec<Box<OpaqueStruct>> {
     vec![Box::new(get_new_struct())]
 }
 
+/// Return array wrapped in a tuple
```
[the remark this answers is not in the document]
Comment (mversic) [reply ffi/tests/ffi_export.rs:183]: yeah, some artefact

File: ffi/src/ir.rs (hunk 1 above)
Comment (mversic) [reply ffi/src/ir.rs:136]: good suggestion

**Review by Erigara — APPROVED** [review]

**Review by Erigara — DISMISSED** [review]

**Review by Erigara — DISMISSED** [review]

**Review by SamHSmith — DISMISSED** [review]
Seems fine to me.

**Review by SamHSmith — APPROVED** [review]

**Review by Erigara — APPROVED** [review]

**Pull request merged**